dali/pipeline/operators/crop/crop.h

```diff
@@ -1,176 +1,188 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_CROP_CROP_H_
+#define DALI_PIPELINE_OPERATORS_CROP_CROP_H_
+
+#include <vector>
+#include <utility>
+
+#include "dali/common.h"
+#include "dali/error_handling.h"
+#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operators/operator.h"
+
+namespace dali {
+
+class CropAttr {
+ protected:
+  explicit inline CropAttr(const OpSpec &spec)
+      : image_type_(spec.GetArgument<DALIImageType>("image_type")),
+        C_(IsColor(image_type_) ? 3 : 1),
+        batch_size_{spec.GetArgument<int>("batch_size")} {
+    if (spec.name() != "Resize") {
+      vector<float> cropArgs = spec.GetRepeatedArgument<float>("crop");
+
+      DALI_ENFORCE(cropArgs[0] >=0,
+        "Crop height must be greater than zero. Received: " + std::to_string(cropArgs[0]));
+      DALI_ENFORCE(cropArgs[1] >=0,
+        "Crop width must be greater than zero. Received: " + std::to_string(cropArgs[0]));
+
+      crop_height_ = std::vector<int>(batch_size_, static_cast<int>(cropArgs[0]));
+      crop_width_ = std::vector<int>(batch_size_, static_cast<int>(cropArgs[1]));
+
+      crop_x_norm_ = std::vector<float>(batch_size_, -1.0f);
+      crop_y_norm_ = std::vector<float>(batch_size_, -1.0f);
+    }
+  }
+
+  std::pair<int, int> SetCropXY(const OpSpec &spec, const ArgumentWorkspace *ws,
+                                const Index dataIdx, int H, int W) {
+    DALI_ENFORCE(H >= crop_height_[dataIdx]);
+    DALI_ENFORCE(W >= crop_width_[dataIdx]);
+
+    auto crop_x_norm =
+        std::vector<float>(batch_size_, spec.GetArgument<float>("crop_pos_x", ws, dataIdx));
+    auto crop_y_norm =
+        std::vector<float>(batch_size_, spec.GetArgument<float>("crop_pos_y", ws, dataIdx));
+
+    DALI_ENFORCE(crop_y_norm[dataIdx] >= 0.f && crop_y_norm[dataIdx] <= 1.f,
+                 "Crop coordinates need to be in range [0.0, 1.0]");
+    DALI_ENFORCE(crop_x_norm[dataIdx] >= 0.f && crop_x_norm[dataIdx] <= 1.f,
+                 "Crop coordinates need to be in range [0.0, 1.0]");
+
+    const int crop_y = crop_y_norm[dataIdx] * (H - crop_height_[dataIdx]);
+    const int crop_x = crop_x_norm[dataIdx] * (W - crop_width_[dataIdx]);
+
+    return std::make_pair(crop_y, crop_x);
+  }
+
+  const vector<Index> CheckShapes(const SampleWorkspace *ws) {
+    const auto &input = ws->Input<CPUBackend>(0);
+
+    DALI_ENFORCE(input.shape().size() == 3,
+                 "Expects 3-dimensional image input.");
+
+    return input.shape();
+  }
+
+  vector<int> crop_height_;
+  vector<int> crop_width_;
+
+  vector<float> crop_x_norm_;

```

Comment:
Shouldn't this be defined in Slice?

Reply:
After my previous set of changes we can get rid completely of crop_x_norm as it is not used anymore. Gpu version can rely only on crop_height and crop_width